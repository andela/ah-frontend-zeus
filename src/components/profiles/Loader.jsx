import React, { Component } from 'react';

class Loader extends Component {
  render() {
    return (
      <div id="loader-container" style={{ display: this.props.display }}>
        <div id="loader" />
      </div>
    );
  }
}

export default Loader;
