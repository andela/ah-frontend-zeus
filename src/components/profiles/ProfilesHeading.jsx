import React, { Component } from 'react';

class ProfilesHeading extends Component {
  render() {
    return (
      <section id="dashboard-page" className="flex-grow-1">
        <section id="profile" className="mt-5 mb-2">
          <div className="container">
            <div className="row">
              <div className="col-md-10 m-auto">
                <h2 className="display-2" style={{ textAlign: 'center' }}>
                  {this.props.heading}
                  <input
                    style={{ display: this.props.displayFilter ? 'block' : 'none' }}
                    onChange={this.props.handleChange}
                    name="display"
                    placeholder="Filter Users..."
                    id="filter-users"
                  />
                </h2>
              </div>
            </div>
          </div>
        </section>
      </section>
    );
  }
}

export default ProfilesHeading;
