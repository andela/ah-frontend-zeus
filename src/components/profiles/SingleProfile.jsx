import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { followUser, unfollowUser } from '../../actions/FollowsActions';

export class SingleProfile extends Component {
  constructor(props) {
    super(props);
    this.followUser = this.followUser.bind(this);
    this.unfollowUser = this.unfollowUser.bind(this);
  }

  followUser() {
    this.props.followUser(this.props.name);
  }

  unfollowUser() {
    this.props.unfollowUser(this.props.name);
  }

  render() {
    return (
      <section id="dashboard-page" className="flex-grow-1">
        <section id="profile" className="mt-5 mb-2">
          <div className="container">
            <div className="row">
              <div className="col-md-10 m-auto">
                <h4 className="display-4">{this.props.name}</h4>
                <div className="d-flex flex-row">
                  <div className="col-md-8">
                    <p className="lead">About: {this.props.about}</p>
                    <p className="lead">Fun Fact: {this.props.funFact}</p>
                    <p>
                      <span>
                        <Link
                          to={'/' + this.props.name + '/followings'}
                          className="text-sm mr-3 text-muted"
                        >
                          <i className="fas fa-users" /> Following
                        </Link>
                      </span>
                      <span>
                        <Link
                          to={'/' + this.props.name + '/followers'}
                          className="text-sm text-muted"
                        >
                          <i className="fas fa-users" /> Followers
                        </Link>
                      </span>
                    </p>
                    {this.props.isFollowing ? (
                      <button
                        onClick={this.unfollowUser}
                        className="btn-sm btn-outline-primary"
                      >
                        Unfollow
                      </button>
                    ) : (
                      <button
                        onClick={this.followUser}
                        className="btn-primary btn-sm"
                      >
                        Follow
                      </button>
                    )}
                  </div>
                  <div className="col-md-2 margin-top">
                    {this.props.image ? (
                      <img
                        className="img-thumbnail rounded-circle"
                        src={this.props.image}
                        alt=""
                      />
                    ) : (
                      <i className="far fa-user-circle fa-8x" />
                    )}
                  </div>
                </div>
              </div>
            </div>
          </div>
        </section>
      </section>
    );
  }
}

export default connect(
  null,
  { followUser, unfollowUser }
)(SingleProfile);
