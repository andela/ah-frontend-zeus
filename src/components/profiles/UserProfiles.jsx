import React, { Component } from 'react';
import { connect } from 'react-redux';
import {
  getUserProfiles,
  getFollowsForCurrentUser,
  getFollowersForCurrentUser,
  getFollowsForAnyUser,
  getFollowersForAnyUser
} from '../../actions/FollowsActions';
import SingleProfile from './SingleProfile';
import ProfilesHeading from './ProfilesHeading';

export class UserProfiles extends Component {
  constructor(props) {
    super(props);
    this.state = { filter: '' };
    this.handleChange = this.handleChange.bind(this);
  }

  handleChange(e) {
    this.setState({ filter: e.target.value.toLowerCase() });
  }

  componentDidMount() {
    this.props.getUserProfiles();
    this.props.getFollowsForCurrentUser();
    this.props.getFollowersForCurrentUser();
    if (this.props.followsFor) {
      this.props.getFollowsForAnyUser(this.props.followsFor);
    } else if (this.props.followersFor) {
      this.props.getFollowersForAnyUser(this.props.followersFor);
    }
  }

  render() {
    let profiles;
    let heading;
    let displayFilter = false;
    if (this.props.followsFor === '_LoggedInUser_') {
      profiles = this.props.authors.filter(author =>
        this.props.follows.includes(author.username)
      );
      heading = profiles.length
        ? 'People you are following'
        : 'You are not following anyone';
    } else if (
      this.props.followsFor === this.props.userWhoseFollowsInfoIsDisplayed
    ) {
      profiles = this.props.authors.filter(author =>
        this.props.followsForAnyUser.includes(author.username)
      );
      heading = profiles.length
        ? 'People ' + this.props.followsFor + ' is following'
        : this.props.followsFor + ' is not following anyone';
    } else if (this.props.followersFor === '_LoggedInUser_') {
      profiles = this.props.authors.filter(author =>
        this.props.followers.includes(author.username)
      );
      heading = profiles.length
        ? 'People following you'
        : 'No one is following you';
    } else if (
      this.props.followersFor === this.props.userWhoseFollowsInfoIsDisplayed
    ) {
      profiles = this.props.authors.filter(author =>
        this.props.followersForAnyUser.includes(author.username)
      );
      heading = profiles.length
        ? 'People following ' + this.props.followersFor
        : 'No one is following ' + this.props.followersFor;
    } else {
      profiles = this.props.authors.filter(
        author => author.username != this.props.username
      );
      profiles = profiles.filter(author =>
        author.username.toLowerCase().startsWith(this.state.filter)
      );
      heading = 'User Profiles';
      displayFilter = true;
    }
    return (
      <React.Fragment>
        <ProfilesHeading
          heading={heading}
          displayFilter={displayFilter}
          handleChange={this.handleChange}
        />
        {profiles.map(author => {
          return (
            <SingleProfile
              key={author.email}
              name={author.username}
              about={author.profile.bio}
              funFact={author.profile['fun_fact']}
              image={author.profile.photo}
              isFollowing={this.props.follows.includes(author.username)}
            />
          );
        })}
      </React.Fragment>
    );
  }
}

function mapStateToProps(state) {
  return {
    authors: state.authors,
    username: state.currentUser.username,
    follows: state.currentUser.followings,
    followers: state.currentUser.followers,
    userWhoseFollowsInfoIsDisplayed:
      state.userWhoseFollowsInfoIsDisplayed.username,
    followsForAnyUser: state.userWhoseFollowsInfoIsDisplayed.followings,
    followersForAnyUser: state.userWhoseFollowsInfoIsDisplayed.followers
  };
}

export default connect(
  mapStateToProps,
  {
    getUserProfiles,
    getFollowsForCurrentUser,
    getFollowersForCurrentUser,
    getFollowsForAnyUser,
    getFollowersForAnyUser
  }
)(UserProfiles);
