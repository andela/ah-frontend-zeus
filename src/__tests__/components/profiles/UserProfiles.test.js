import React from 'react';
import Enzyme from 'enzyme';
import EnzymeAdapter from 'enzyme-adapter-react-16';
import { shallow } from 'enzyme';
import { UserProfiles } from '../../../components/profiles/UserProfiles';
import ProfilesHeading from '../../../components/profiles/ProfilesHeading';
import { SingleProfile } from '../../../components/profiles/SingleProfile';

Enzyme.configure({ adapter: new EnzymeAdapter() });

function setup() {
  const props = {
    getUserProfiles: jest.fn(),
    getFollowsForCurrentUser: jest.fn(),
    getFollowersForCurrentUser: jest.fn(),
    username: 'LoggedInUser',
    follows: ['one', 'two'],
    followsFor: '',
    followersFor: '',
    authors: [
      {
        username: 'Username',
        email: 'username@host.com',
        profile: { bio: 'bio', fun_fact: 'fun_fact', photo: 'photo_url' }
      }
    ]
  };

  const wrapper = shallow(<UserProfiles {...props} />);
  return { wrapper, props };
}

describe('<UserProfiles />', () => {
  it('contains a <ProfilesHeading />', () => {
    const { wrapper } = setup();
    expect(wrapper.find(ProfilesHeading)).toHaveLength(1);
  });

  it('matches snapshot', () => {
    const { wrapper } = setup();
    expect(wrapper.find(SingleProfile)).toExist;
    expect(wrapper).toMatchSnapshot();
  });

  it('calls props methods when mounted', () => {
    const { props } = setup();
    expect(props.getUserProfiles).toHaveBeenCalled();
    expect(props.getFollowsForCurrentUser).toHaveBeenCalled();
    expect(props.getFollowersForCurrentUser).toHaveBeenCalled();
  });
});
