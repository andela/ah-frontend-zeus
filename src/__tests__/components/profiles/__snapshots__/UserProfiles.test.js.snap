// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserProfiles /> matches snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <UserProfiles
    authors={
      Array [
        Object {
          "email": "username@host.com",
          "profile": Object {
            "bio": "bio",
            "fun_fact": "fun_fact",
            "photo": "photo_url",
          },
          "username": "Username",
        },
      ]
    }
    followersFor=""
    follows={
      Array [
        "one",
        "two",
      ]
    }
    followsFor=""
    getFollowersForCurrentUser={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    getFollowsForCurrentUser={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    getUserProfiles={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    username="LoggedInUser"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <ProfilesHeading
          displayFilter={true}
          handleChange={[Function]}
          heading="User Profiles"
        />,
        Array [
          <Connect(SingleProfile)
            about="bio"
            funFact="fun_fact"
            image="photo_url"
            isFollowing={false}
            name="Username"
          />,
        ],
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "displayFilter": true,
          "handleChange": [Function],
          "heading": "User Profiles",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "username@host.com",
        "nodeType": "class",
        "props": Object {
          "about": "bio",
          "funFact": "fun_fact",
          "image": "photo_url",
          "isFollowing": false,
          "name": "Username",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <ProfilesHeading
            displayFilter={true}
            handleChange={[Function]}
            heading="User Profiles"
          />,
          Array [
            <Connect(SingleProfile)
              about="bio"
              funFact="fun_fact"
              image="photo_url"
              isFollowing={false}
              name="Username"
            />,
          ],
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "displayFilter": true,
            "handleChange": [Function],
            "heading": "User Profiles",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "username@host.com",
          "nodeType": "class",
          "props": Object {
            "about": "bio",
            "funFact": "fun_fact",
            "image": "photo_url",
            "isFollowing": false,
            "name": "Username",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
